⭐ Create E2E test setup in Go

imilchev opened the pull request:
closes #261

Signed-off-by: Ivan Milchev <ivan@mondoo.com>

imilchev commented:
I added the initial framework for the E2E tests in Go. I still have the old E2E enabled since I haven't re-implemented the tests.

@joelddiaz can we merge this PR so the framework is already in `main` and then I will make a separate PR for re-implementing the tests and removing the old E2E. In this way we can also see the tests being run as I work on them in the new PR.

joelddiaz commented on tests/integration/test_cluster.go:
```diff
@@ -0,0 +1,45 @@
+		T:               t,
+	}
+
+	if err := cluster.InstallOperator(); err != nil {
```
Should we have some code to clear up any previous resources that might be in the mondoo-operator namespace? Basically, a full cleanup before we start?
When trying this locally, I had a previous installation (that was separate from anything related to this PR), and I would get a bunch of errors as the install was happening.

imilchev replied:
i would not like that because maybe you have some local stuff that you don't want magically deleted. That's why I didn't do it

joelddiaz replied:
Fair enough. Just wanted to make sure the behavior was intentional.

joelddiaz commented on tests/integration/e2e_example_test.go:
```diff
@@ -19,6 +20,12 @@ func (s *E2ESuite) TearDownSuite() {
 
+func (s *E2ESuite) AfterTest(suiteName, testName string) {
+	if !s.T().Failed() {
+		s.testCluster.GatherAllMondooLogs(testName, installer.MondooNamespace)
```
This is fantastic.

joelddiaz commented on .github/workflows/integration.yaml:
```diff
@@ -0,0 +1,37 @@
+  pull_request:
+  push:
+    paths-ignore:
+      - 'docs/**'
```
Can safely ignore any markdown at the root of the repo as well

imilchev replied:
i deleted this file already. Changed the structure of the tests in the latest commits

joelddiaz reviewed (commented):
Being able to run this straight from VSCode is great!

We should put some comments in the files that were brought over from the rook repo (so future people can have an idea where these files came from).

imilchev commented on tests/framework/installer/installer.go:
```diff
@@ -0,0 +1,85 @@
+
+func (i *MondooInstaller) UninstallOperator() error {
+	// If the operator has not been installed do nothing
+	if !i.isInstalled {
```
that is correct, I already realized that while i was working on a real test

imilchev commented on tests/integration/test_cluster.go:
```diff
@@ -0,0 +1,45 @@
+		}
+		t().Fail()
+		cluster.UninstallOperator()
+		t().FailNow()
```
does it actually try to execute `AfterTest()` because I think it shouldn't because of the `FailNow()`?

imilchev commented on tests/integration/mondoo_installation_test.go:
```diff
@@ -29,12 +36,64 @@ func (s *MondooInstallationSuite) AfterTest(suiteName, testName string) {
+func (s *MondooInstallationSuite) testMondooInstallation() {
+	zap.S().Info("Create an audit config that enables only workloads scanning.")
+	auditConfig := utils.DefaultAuditConfig(s.testCluster.Settings.Namespace, true, false, false)
+	s.NoError(s.testCluster.K8sHelper.Clientset.Create(s.ctx, &auditConfig))
```
Agreed. I documented the places where it can go wrong but I didn't add comments to every single place an error can occur as I think that is unnecessary.

imilchev commented on tests/framework/utils/k8s_helper.go:
```diff
@@ -28,10 +28,12 @@ const (
+	DeleteArgs                        = []string{"delete", "-f"}
+	DeleteArgsIgnoreNotFound          = []string{"delete", "--ignore-not-found=true", "-f"}
+	DeleteFromStdinArgs               = append(DeleteArgs, "-")
+	DeleteIngoreNotFoundFromStdinArgs = append(DeleteArgs, "-")
```
I completely overlooked that one. Good catch!

joelddiaz reviewed (commented):
This looks great.

One last request. Can you put some comments at the top of the files copied over from the rook repo to indicate where they were copied from, and preserve their original copyrights?

imilchev commented:
> This looks great.
> 
> One last request. Can you put some comments at the top of the files copied over from the rook repo to indicate where they were copied from, and preserve their original copyrights?

I can add a link to their test folder but the code is not copy-pasted. I adjusted it a bit and also deleted a lot from it that we do not actually need. I will just mention that the idea is taken from there and the code has been adapted

joelddiaz commented:
> > This looks great.
> > One last request. Can you put some comments at the top of the files copied over from the rook repo to indicate where they were copied from, and preserve their original copyrights?
> 
> I can add a link to their test folder but the code is not copy-pasted. I adjusted it a bit and also deleted a lot from it that we do not actually need. I will just mention that the idea is taken from there and the code has been adapted

After chatting it over with @chris-rock , I think our best bet is to preserve the Rook copyright, and to also go ahead and add `Copyright 2022 Mondoo, Inc.` to the files where you made substantial additions/changes.

imilchev commented on tests/framework/utils/k8s_helper.go:
```diff
@@ -0,0 +1,347 @@
+/*
```
this file is currently 50:50 between Rook code and our code. It is highly likely that over time it will change more and more to something that we need. Not sure what to do with this specific copyright

joelddiaz commented on tests/framework/utils/k8s_helper.go:
```diff
@@ -0,0 +1,363 @@
+/*
+Copyright 2022 Mondoo, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+Copyright 2016 The Rook Authors. All rights reserved.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+	http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
```
Since it is the same license, I think you can just collapse this all into one section
```suggestion
/*
Copyright 2016 The Rook Authors. All rights reserved.
Copyright 2022 Mondoo, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
```

joelddiaz reviewed (commented):
👍 LGTM
